<a href="https://colab.research.google.com/github/jrivest2/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module1/Justin_Rivest_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [39]:
### YOUR WORK HERE
import pandas as pd
import numpy as np
from scipy import stats

column_names = ['symboling', 'normalized-losses', 'make', 'fuel-type', 
                'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 
                'engine-location', 'wheel-base', 'length', 'width', 'height', 
                'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 
                'fuel-system', 'bore', 'stroke', 'compression-ratio', 
                'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']

df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data',
                 names=column_names)
df

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0            3                 ?  alfa-romero  ...       21          27  13495
1            3                 ?  alfa-romero  ...       21          27  16500
2            1                 ?  alfa-romero  ...       19          26  16500
3            2               164         audi  ...       24          30  13950
4            2               164         audi  ...       18          22  17450
..         ...               ...          ...  ...      ...         ...    ...
200         -1                95        volvo  ...       23          28  16845
201         -1                95        volvo  ...       19          25  19045
202         -1                95        volvo  ...       18          23  21485
203         -1                95        volvo  ...       26          27  22470
204         -1                95        volvo  ...       19          25  22625

[205 rows x 26 columns]

In [42]:
df = df.replace('?',np.NaN)
df

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0            3               NaN  alfa-romero  ...       21          27  13495
1            3               NaN  alfa-romero  ...       21          27  16500
2            1               NaN  alfa-romero  ...       19          26  16500
3            2               164         audi  ...       24          30  13950
4            2               164         audi  ...       18          22  17450
..         ...               ...          ...  ...      ...         ...    ...
200         -1                95        volvo  ...       23          28  16845
201         -1                95        volvo  ...       19          25  19045
202         -1                95        volvo  ...       18          23  21485
203         -1                95        volvo  ...       26          27  22470
204         -1                95        volvo  ...       19          25  22625

[205 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. 

---

Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [47]:
### YOUR WORK HERE
sample = df.sample(10,random_state=30)
sample

symboling normalized-losses           make  ... city-mpg highway-mpg  price
121          1               154       plymouth  ...       31          38   6692
176         -1                65         toyota  ...       27          32  10898
24           1               148          dodge  ...       31          38   6229
88          -1               137     mitsubishi  ...       23          30   9279
72           3               142  mercedes-benz  ...       16          18  35056
103          0               108         nissan  ...       19          25  13499
26           1               148          dodge  ...       31          38   7609
20           0                81      chevrolet  ...       38          43   6575
155          0                91         toyota  ...       27          32   8778
185          2                94     volkswagen  ...       27          34   8195

[10 rows x 26 columns]

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

###Null Hypothesis:

The cars the salesman sells typically have a fuel efficiency of about 35 miles per gallon on the highway.

###Alternative Hypothesis:

The cars the salesman sells **do not** typically have a fuel efficiency of about 35 miles per gallon on the highway.

### Confidence Level: 95%

In [49]:
### YOUR WORK HERE
stats.ttest_1samp(sample['highway-mpg'],35,nan_policy='omit')

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

###Conclusion:
Based on a t-statistic of -0.96, and a p-value of 0.36, I (reject/**fail to reject**) the null hypothesis that the cars the salesman sells typically have a fuel efficiency of about 35 miles per gallon on the highway.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [53]:
### YOUR WORK HERE
# Sample 100 cars
new_sample = df.sample(100,random_state=30)
new_sample

symboling normalized-losses           make  ... city-mpg highway-mpg  price
121          1               154       plymouth  ...       31          38   6692
176         -1                65         toyota  ...       27          32  10898
24           1               148          dodge  ...       31          38   6229
88          -1               137     mitsubishi  ...       23          30   9279
72           3               142  mercedes-benz  ...       16          18  35056
..         ...               ...            ...  ...      ...         ...    ...
189          3               NaN     volkswagen  ...       24          29  11595
192          0               NaN     volkswagen  ...       33          38  13845
171          2               134         toyota  ...       24          30  11549
203         -1                95          volvo  ...       26          27  22470
47           0               145         jaguar  ...       15          19  32250

[100 rows x 26 columns]

###Null Hypothesis:

The cars the salesman sells typically have a fuel efficiency of about 35 miles per gallon on the highway.

###Alternative Hypothesis:

The cars the salesman sells **do not** typically have a fuel efficiency of about 35 miles per gallon on the highway.

###Confidence Level: 95%

In [54]:
stats.ttest_1samp(new_sample['highway-mpg'],35,nan_policy='omit')

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

###Conclusion:

Based on a t-statistic of -6.52, and a p-value of 0.0000000036, I (**reject**/fail to reject) the null hypothesis that the cars the salesman sells typically have a fuel efficiency of about 35 miles per gallon on the highway.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

It's because using a larger sample size allowed me to assess the population with greater accuracy.

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [2]:
### YOUR WORK HERE
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-09-07 21:07:22--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data.1’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.05s   

2020-09-07 21:07:22 (368 KB/s) - ‘house-votes-84.data.1’ saved [18171/18171]



In [20]:
# Read the dataset in from UCI, you'll need to provide a list of column headers
column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']
df = pd.read_csv('house-votes-84.data', names=column_headers)
print(df.shape)
df

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0    republican                   n             y  ...     y         n            y
1    republican                   n             y  ...     y         n            ?
2      democrat                   ?             y  ...     y         n            n
3      democrat                   n             y  ...     n         n            y
4      democrat                   y             y  ...     y         y            y
..          ...                 ...           ...  ...   ...       ...          ...
430  republican                   n             n  ...     y         n            y
431    democrat                   n             n  ...     n         n            y
432  republican                   n             ?  ...     y         n            y
433  republican                   n             n  ...     y         n            y
434  republican                   n             y  ...     y         ?            n

[435 rows x 17 columns]

In [10]:
# Encode "yes" votes as 1 and "no" votes as 0.
df = df.replace({'y':1,'n':0,'?':np.NaN})
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [11]:
# Use dataframe filtering to split the dataframe into two new dataframes based
#  on party. Hold all republicans in one dataframe and all democrats in the
#  other. These will be our two different "samples."
rep = df[df['party'] == 'republican']
dem = df[df['party'] == 'democrat']

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [ ]:
### YOUR WORK HERE

###Null Hypothesis:

Republican support for the budget bill is equal to democrat levels of support for the immigration bill.

###Alternative Hypothesis:

Republican support for the budget bill is not equal to democrat levels of support for the immigration bill.

###Confidence Level: 99%

In [16]:
# Find an issue that democrats support more than republicans with p < 0.01
#  (significant at the 99% level).
stats.ttest_ind(dem['budget'],rep['budget'],nan_policy='omit')

Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

###Conclusion:

Based on a t-statistic of 23.213, and a p-value of 0.000...0027, I (**reject**/fail to reject) the null hypothesis that republican levels of support are equal to democrat levels of support for the budget bill. Also, based on a t-statistic of 23.213, the democrat party supports this bill far more.

###Null Hypothesis:

Republican support for the religious-groups bill is equal to democrat levels of support for the religious-groups bill.

###Alternative Hypothesis:

Republican support for the religious-groups bill is not equal to democrat levels of support for the religious-groups bill.

###Confidence Level: 99%

In [34]:
# Find an issue that republicans support more than democrats with p < 0.01
#  (significant at the 99% level).
stats.ttest_ind(dem['south-africa'], rep['south-africa'], nan_policy='omit')

Ttest_indResult(statistic=6.849454815841208, pvalue=3.652674361672226e-11)

###Conclusion:

Based on a t-statistic of -9.74, and a p-value of 0.000...0024, I (**reject**/fail to reject) the null hypothesis that republican levels of support are equal to democrat levels of support for the religious-groups bill. Also based on a t-statistic of -1.74, the republicans support the religious-groups bill more.

###Null Hypothesis:

Republican levels of support for the water-project bill are equal to democrat levels of support for the water-project bill.

###Alternative Hypothesis:

Republican levels of support for the water-project bill are not equal to democrat levels of support for the water-project bill.

###Confidence Level: 90%

In [38]:
# Find an issue where the difference between republicans and democrats has
#  p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a
#  difference the two sample means)
stats.ttest_ind(dem['water-project'],rep['water-project'],nan_policy='omit')

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

###Conclusion:

Based on a t-statistic of -0.089, and a p-value of 0.93, I (reject/**fail to reject**) the null hypothesis that republican levels of support are equal to democrat levels of support for the water-project bill.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
### YOUR WORK HERE



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE